qforce/qm/orca.py

```diff
@@ -0,0 +1,624 @@
+                charges.append(float(charge))
+        # atom_id is zero-based index
+        return int(atom_id) + 1, charges
+
+    @staticmethod
+    def _read_orca_coords(coord_text):
+        """ Read the optimised coordinate string.
+
+        The string is in the format of the standard xyz file.
+
+        Parameters
+        ----------
+        coord_text : string
+            The content of the xyz file file.
+
+        Returns
+        -------
+        n_atoms : int
+            The number of atoms in the molecule.
+        elements : array
+            A np.array of integer of the atomic number of the atoms.
+        coords : array
+            An array of float of the shape (n_atoms, 3).
+        """
+        lines = coord_text.split('\n')
+        n_atoms = int(lines[0])
+        coords = np.empty((n_atoms, 3))
+        elements = []
+        for i, line in enumerate(lines[2: 2+n_atoms]):
+            element, x, y, z = line.split()
+            elements.append(ATOM_SYM.index(element))
+            coords[i, :] = (x, y, z)
+        return n_atoms, np.array(elements), coords
+
+    @staticmethod
+    def _read_orca_xyz(coord_file):
+        """ Read the optimised coordinate xyz file.
+
+        For ORCA jobs, the optimised geometry will be stored as a file
+        with the extension xyz.
+
+        Parameters
+        ----------
+        coord_file : string
+            The name of the coordinates file.
+
+        Returns
+        -------
+        n_atoms : int
+            The number of atoms in the molecule.
+        elements : array
+            A np.array of integer of the atomic number of the atoms.
+        coords : array
+            An array of float of the shape (n_atoms, 3).
+        """
+        with open(coord_file, 'r') as f:
+            coord_text = f.read()
+        return ReadORCA._read_orca_coords(coord_text)
+
+    @staticmethod
+    def _read_orca_allxyz(coord_file):
+        """ Read the optimised coordinate allxyz file.
+
+        For ORCA jobs, the scan optimised geometry will be stored as a file
+        with the extension allxyz.
+
+        Parameters
+        ----------
+        coord_file : string
+            The name of the coordinates file.
+
+        Returns
+        -------
+        n_atoms : int
+            The number of atoms in the molecule.
+        elements : array
+            A np.array of integer of the atomic number of the atoms.
+        coords : list
+            A list of array of float. The list has the length of the number
+            of steps and the array has the shape of (n_atoms, 3).
+        """
+        with open(coord_file, 'r') as f:
+            coord_text = f.read()
+        coords = []
+        for text in coord_text.split('>\n'):
+            n_atoms, elements, coord = ReadORCA._read_orca_coords(text)
+            coords.append(coord)
+        return n_atoms, elements, coords
+
+    @staticmethod
+    def _read_orca_nbo_analysis(out_file, n_atoms):
```

Comment:
could you not use the inherited _read_nbo_analysis function in qm_base? Looks like a very similar code

Reply:
I have thought about it but I need to change some subtle things to make it work with ORCA.

https://github.com/selimsami/qforce/blob/e51178fe33f279a3bb08c1a3383ab59ecd67fc83/qforce/qm/qm_base.py#L89
Which is `atom = int(line[16:20])` in orca.

https://github.com/selimsami/qforce/blob/e51178fe33f279a3bb08c1a3383ab59ecd67fc83/qforce/qm/qm_base.py#L90
Which is `atom = int(line[32:39])` in orca.

I felt that instead of passing these to a function which then calls the _read_nbo_analysis, it might be more readable to use separate functions.

Reply:
The difference is caused by NBO versions or ORCA? If former, we could decide the indices based on the NBO version. Would be a neater approach as one could encounter any NBO version in any QM software.

I am, however, okay with raising this as an issue an moving forward with merging this PR after the method sorting is updated.

Reply:
Issue #25